src/ralph/attachments/models.py

```diff
@@ -0,0 +1,91 @@
+import os
+import string
+from uuid import uuid4
+
+from django.conf import settings
+from django.contrib.contenttypes import generic
+from django.contrib.contenttypes.models import ContentType
+from django.db import models
+
+from ralph.lib.mixins.models import TimeStampMixin
+
+
+def get_file_path(instance, filename):
+    ext = os.path.splitext(filename)[1]
+    name = ''.join([str(uuid4()), ext])
+    return os.path.join('attachments', name[:1], name[1:2], name)
+
+
+class AttachmentManager(models.Manager):
+    def get_attachments_for_object(self, obj):
+        content_type = ContentType.objects.get_for_model(obj)
+        return Attachment.objects.filter(
+            items__content_type=content_type,
+            items__object_id=obj.id,
+        )
+
+
+class AttachmentItemManager(models.Manager):
+    def get_items_for_object(self, obj):
+        content_type = ContentType.objects.get_for_model(obj)
+        return self.get_queryset().select_related('attachment').filter(
+            object_id=obj.pk,
+            content_type=content_type,
+        )
+
+    def attach(self, pk, content_type, attachments):
+        new_items = []
+        for attachment in attachments:
+            new_items.append(self.model(
+                content_type=content_type,
+                object_id=pk,
+                attachment=attachment,
+            ))
+        if new_items:
+            self.bulk_create(new_items)
+
+    def dettach(self, pk, content_type, attachments):
+        self.filter(attachment__in=attachments).delete()
+
+    def refresh(self, obj, new_objects=None, deleted_objects=None):
+        content_type = ContentType.objects.get_for_model(obj)
+        if new_objects:
+            self.attach(obj.pk, content_type, new_objects)
+        if deleted_objects:
+            self.dettach(obj.pk, content_type, deleted_objects)
+
+
+class Attachment(TimeStampMixin, models.Model):
+    original_filename = models.CharField(
+        max_length=255,
+        unique=False,
+    )
+    file = models.FileField(upload_to=get_file_path, max_length=255)
+    mime_type = models.CharField(
+        max_length=100,
+        unique=False,
+        default='application/octet-stream',
+    )
+    description = models.TextField(blank=True)
+    uploaded_by = models.ForeignKey(settings.AUTH_USER_MODEL)
+
+    objects = AttachmentManager()
+
+    @staticmethod
+    def _safe_filename(filename):
+        allowed = '-_.() ' + string.ascii_letters + string.digits
```

Comment:
what if there will be some non-ascii letter in attachment name (ex. polish letter)? we're storing this file under our own name, so we could allow every (escaped of course) filename IMO.